Add `--semaphore` flag to enable interaction with GHC Job Server…

mpickering opened the pull request:
… (-jsem)

This is a WIP patch which allows testing GHC jobserver (-jsem) using cabal. When `--semaphore` is passed to `cabal-install` then cabal will create a semaphore with the number of capabilities suggested by `-j` and then use this semaphore to mediate parallelism between GHC processes.

sheaf commented on Cabal/src/Distribution/Simple/Build.hs:
```diff
@@ -717,15 +740,15 @@ buildLib verbosity numJobs pkg_descr lbi lib clbi =
       _   -> die' verbosity "Building is not supported with this compiler."
 
-buildExe :: Verbosity -> Flag (Maybe Int)
+buildExe :: Verbosity -> Flag (ParStrat)
```
```suggestion
buildExe :: Verbosity -> Flag ParStrat
```

sheaf commented on Cabal/src/Distribution/Simple/GHC.hs:
```diff
@@ -1002,14 +1003,14 @@ startInterpreter verbosity progdb comp platform packageDBs = do
 replFLib
   :: ReplOptions             -> Verbosity
-  -> Flag (Maybe Int)  -> PackageDescription
+  -> Flag (ParStrat)  -> PackageDescription
```
```suggestion
  -> Flag ParStrat  -> PackageDescription
```

sheaf commented on Cabal/src/Distribution/Simple/GHC.hs:
```diff
@@ -1313,7 +1314,7 @@ replNoLoad replFlags l
 
 -- | Generic build function. See comment for 'GBuildMode'.
-gbuild :: Verbosity          -> Flag (Maybe Int)
+gbuild :: Verbosity          -> Flag (ParStrat)
```
```suggestion
gbuild :: Verbosity          -> Flag ParStrat
```

sheaf commented on Cabal/src/Distribution/Simple/GHCJS.hs:
```diff
@@ -740,14 +741,14 @@ startInterpreter verbosity progdb comp platform packageDBs = do
 -- | Build a foreign library
 buildFLib
-  :: Verbosity          -> Flag (Maybe Int)
+  :: Verbosity          -> Flag (ParStrat)
```
```suggestion
  :: Verbosity          -> Flag ParStrat
```

sheaf commented on Cabal/src/Distribution/Simple/GHCJS.hs:
```diff
@@ -756,14 +757,14 @@ replFLib replFlags  v njobs pkg lbi =
 replExe
   :: [String]                -> Verbosity
-  -> Flag (Maybe Int)  -> PackageDescription
+  -> Flag (ParStrat)  -> PackageDescription
```
```suggestion
  -> Flag ParStrat  -> PackageDescription
```

sheaf commented on Cabal/src/Distribution/Simple/GHCJS.hs:
```diff
@@ -1044,7 +1045,7 @@ isHaskell :: FilePath -> Bool
 
 -- | Generic build function. See comment for 'GBuildMode'.
-gbuild :: Verbosity          -> Flag (Maybe Int)
+gbuild :: Verbosity          -> Flag (ParStrat)
```
```suggestion
gbuild :: Verbosity          -> Flag ParStrat
```

sheaf commented on Cabal/src/Distribution/Types/ParStrat.hs:
```diff
@@ -0,0 +1,9 @@
+module Distribution.Types.ParStrat where
+
+data ParStrat = UseSem String Int | Old (Maybe Int) | Serial
```
I think you should rename the `Old` constructor and add some documentation, e.g.:

```suggestion
-- | How to control parallelism, e.g. a fixed number of jobs or by using a system semaphore.
data ParStrat
  -- | Compile in parallel with the given number of jobs (`-jN` or `-j`).
  = NumJobs (Maybe Int)
  -- | `--semaphore`: use a system semaphore to control parallelism.
  | UseSem String Int
  -- | No parallelism (neither `-j` nor `--semaphore`).
  | Serial
```

sheaf commented on cabal-install/src/Distribution/Client/JobControl.hs:
```diff
@@ -156,6 +166,61 @@ readAllTChan qvar = go []
+        job <- atomically $ readTChan inqVar
+        waitOnSemaphore sem
+        void $ forkIO $ do
+          -- bracket
```
bracket

sheaf commented on doc/cabal-project.rst:
```diff
@@ -269,6 +269,19 @@ package, and thus apply globally:
+
+    :default: False
+
+    When combined with ``jobs``, this option instructs cabal to create a
+    semaphore with the number of slots specified by ``jobs``. This semaphore is
+    then passed to GHC using the `-jsem` option to make use of GHC's ability to
+    control parrelism using a semaphore.
```
```suggestion
    This option instructs cabal to control parallelism by creating a new system semaphore,
    whose number of tokens is specified by ``--jobs`` (or ``-j``).
    This semaphore is passed to GHC, which allows it to use any leftover parallelism
    that ``cabal-install`` is not using.
    
    Requires ``ghc >= 9.8``.
```

Question: what happens if no ``--jobs`` argument is passed? Do we use `$ncpus`, or throw an error?

mpickering replied:
By default `jobs: $ncpus`, but if by default the flag wasn't set then we should create a semaphore with one slot which would just behave like `-j1` normally would.

mpickering commented on cabal-install/src/Distribution/Client/JobControl.hs:
```diff
@@ -156,6 +166,61 @@ readAllTChan qvar = go []
+          res <- try job `finally` releaseSemaphore sem 1
+          atomically $ writeTChan outqVar res
+        -- Wait for spawned process to take some capabilities if it wants some
+        threadDelay 1_000_000
```
I don't think so.. I didn't try any different values. I think just something non-zero is good and not so small to prevent thrashing.

mpickering commented:
Updated based on @sheaf s comments now

* Made the semaphore name fresh, so that you can run multiple cabal processes at once.
* Parametrised the ParStrat type so it can be used by cabal-install and Cabal
* Fixed the logic taking/returning semaphores to correctly handle async exceptions (thanks @sheaf for discussing this with me)

mpickering commented:
Testing this feature will have to wait because there isn't a released compiler yet which supports `-jsem`.

sheaf commented on changelog.d/pr-8557:
```diff
@@ -0,0 +1,20 @@
+control how much parellism it uses in coordation with other simultaneously running
+processes.
+
+This feature requires ghc-9.8 in order to use as this is the first version of GHC
+which implements the protocol.
```
```suggestion
This feature requires ghc-9.8 in order to use, as this is the first version of GHC
which implements the protocol.
```

sheaf commented on changelog.d/pr-8557:
```diff
@@ -0,0 +1,20 @@
+This feature requires ghc-9.8 in order to use as this is the first version of GHC
+which implements the protocol.
+
+The protocol is specified by [GHC Proposal #123]().
```
```suggestion
The protocol is specified by [GHC Proposal #540]().
```

The proposal number should be 540, but it hasn't been merged yet although it's been accepted.

sheaf commented on changelog.d/pr-8557-2:
```diff
@@ -0,0 +1,16 @@
+manage the semaphore according to the GHC Jobserver Protocol.
+
+This low level interface is intended to be called by a high-level tool such as
+`cabal-install` which can create and manage the semaphore appropiately.
```
```suggestion
`cabal-install` which can create and manage the semaphore appropriately.
```

sheaf commented on changelog.d/pr-8557-2:
```diff
@@ -0,0 +1,16 @@
+This low level interface is intended to be called by a high-level tool such as
+`cabal-install` which can create and manage the semaphore appropiately.
+
+The protocol is specified by [GHC Proposal #123]().
```
TODO.

sheaf commented on changelog.d/pr-8557:
```diff
@@ -0,0 +1,20 @@
+This feature requires ghc-9.8 in order to use, as this is the first version of GHC
+which implements the protocol.
+
+The protocol is specified by [GHC Proposal #540]().
```
```suggestion
The protocol is specified by [GHC Proposal #540](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0540-jsem.rst).
```

sheaf commented on changelog.d/pr-8557-2:
```diff
@@ -0,0 +1,16 @@
+This low level interface is intended to be called by a high-level tool such as
+`cabal-install` which can create and manage the semaphore appropriately.
+
+The protocol is specified by [GHC Proposal #123]().
```
```suggestion
The protocol is specified by [GHC Proposal #540](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0540-jsem.rst).
```

fgaz commented:
This will close #976, right?

fgaz commented on cabal-install/src/Distribution/Client/Setup.hs:
```diff
@@ -1946,6 +1947,11 @@ installOptions showOrParseArgs =
+      , option [] ["semaphore"]
+          "Use a semaphore so GHC can compile components in parallel"
+          installUseSemaphore (\v flags -> flags { installUseSemaphore = v })
+          trueArg
```
maybe this should be a `boolOpt`, so it can be enabled **or disabled** in the global or project config

mpickering replied:
I implemented it now so that if you use an older compiler and pass `--sempahore` then you get a warning about falling back to old behaviour but things still work. That way it can be enabled globally without too much hassle.

mpickering commented:
> This will close #976, right?

It seems related to #976 but the idea there could still be implemented.

mpickering commented:
As a data point, using this feature to compile `exe:cabal` in the cabal repo halved the build time.

mpickering commented:
CI is passing now :)

Kleidukos commented on changelog.d/pr-8557:
```diff
@@ -0,0 +1,20 @@
+control how much paralellism it uses, coordinating with other simultaneously running
+processes.
+
+This feature requires ghc-9.8 in order to use, as this is the first version of GHC
```
non-blocking: Shouldn't we always mention GHC 9.7? Because I fear that people may not find anything related to 9.8 if they look it up in the code base.

mpickering replied:
I don't know what is best, it seems that most users won't know what ghc-9.7 is and won't be able to install it.

Kleidukos replied:
Alright, let's leave it like this then. :)

Kleidukos commented on cabal-install/src/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -590,49 +593,58 @@ rebuildTargets verbosity
+         InstallPlan.execute jobControl keepGoing
+                             (BuildFailure Nothing . DependentFailed . packageId)
+                             installPlan $ \pkg ->
+           --TODO: review exception handling
```
haha indeed, that should probably be a ticket in itself. I'll open one.

Kleidukos reviewed (approved):
Nothing blocking on my side, thanks a bunch!

ulysses4ever commented on Cabal/src/Distribution/Simple/Build.hs:
```diff
@@ -146,6 +158,15 @@ build pkg_descr lbi flags suffixes = do
+    unless (jsemSupported comp || not (isJust (flagToMaybe (buildUseSemaphore flags))) ) $
+        die' verbosity $
+              "Your compiler does not support the -jsem flag. "
+           ++ "To use this feature you must use GHC 9.7 or later."
```
Should this say 9.8 for consistency with what the changelog and documentation (parts of this same PR) say?

mpickering replied:
Done.

ulysses4ever commented on Cabal/src/Distribution/Simple/Build.hs:
```diff
@@ -136,7 +138,17 @@ build pkg_descr lbi flags suffixes = do
+    par_strat <- toFlag <$> case buildUseSemaphore flags of
+                  Flag sem_name -> case buildNumJobs flags of
+                                    Flag {} -> do
+                                      warn verbosity $ "Ignoring -j due to -semaphore"
```
```suggestion
                                      warn verbosity $ "Ignoring -j due to --semaphore"
```

mpickering replied:
Done.

ulysses4ever commented on cabal-install/src/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -590,49 +593,58 @@ rebuildTargets verbosity
                              --TODO: [code cleanup] eliminate setup exe cache
 
-    debug verbosity $
+    info verbosity $
```
I'm not sure why the verbosity level had to be bumped.

ulysses4ever commented:
> Testing this feature will have to wait because there isn't a released compiler yet which supports -jsem.

It would be desirable to write a test guarded by the condition `ghcVersion >= 9.7` and run it locally with GHC HEAD. CI will start running it only when a new (pre)release appears, but that's much better than no test whatsoever.

Kleidukos commented:
@mpickering I'll take care of rebasing your PR and fixing the conflicts, if you will allow me.

mpickering commented:
@Kleidukos If you could rebase that would be really good. I added a test and tested it worked locally with a HEAD compiler. I also responded to all the review comments.

Kleidukos commented:
@mpickering certainly!

ulysses4ever commented:
This has been lingering for longer than it should. We really should look into it. The current CI failures are a bit surprising though. But probably nothing serious.

Kleidukos commented:
@sheaf @fgaz ok I believe we're good now but I'd highly appreciate a last review from you so that we can conclude this saga. :)

ffaf1 commented on Cabal/src/Distribution/Simple/Build.hs:
```diff
@@ -162,6 +175,14 @@ build pkg_descr lbi flags suffixes = do
     distPref = fromFlag (buildDistPref flags)
     verbosity = fromFlag (buildVerbosity flags)
 
+checkBuildProblems
```
Could be handy for reviewers to have this function commented.

ffaf1 commented on Cabal/src/Distribution/Simple/Build.hs:
```diff
@@ -162,6 +175,14 @@ build pkg_descr lbi flags suffixes = do
+checkBuildProblems
+  :: Verbosity -> Compiler -> BuildFlags -> IO ()
+checkBuildProblems verbosity comp flags = do
+  unless (jsemSupported comp || not (isJust (flagToMaybe (buildUseSemaphore flags)))) $
```
What is the second condition? Not having --semaphore? It seems not explained in the error message.

Kleidukos replied:
Basically this translates to "Unless the compiler supports semaphores, or the semaphore flag has not been passed, throw an error".

ffaf1 replied:
correct, I would have thought a `when` expressed the warning in a better way, but that is fine.

ffaf1 commented on Cabal/src/Distribution/Simple/Build.hs:
```diff
@@ -145,10 +147,21 @@ build pkg_descr lbi flags suffixes = do
+        Flag sem_name -> case buildNumJobs flags of
+          Flag{} -> do
+            warn verbosity $ "Ignoring -j due to --semaphore"
+            return $ UseSem sem_name
```
`UseSem sem_name` in both branches, maybe it can be put out of the `case`?

Kleidukos replied:
We could probably with `flagToMaybe` and `when` but it would probably compile down to cases in the end :shrug:

ffaf1 commented on Cabal/src/Distribution/Simple/Compiler.hs:
```diff
@@ -363,6 +364,14 @@ unitIdSupported = ghcSupported "Uses unit IDs"
+jsemSupported :: Compiler -> Bool
+jsemSupported comp = case compilerFlavor comp of
+  GHC -> v >= mkVersion [9, 7]
+  _ -> False
```
Never a fan of catch-alls, they can bite us in the future (even though in this case it is low risk).

Kleidukos replied:
With the consolidation of backend behinds GHC, I have no fears of that one specifically.

ffaf1 commented on Cabal/src/Distribution/Simple/Compiler.hs:
```diff
@@ -363,6 +364,14 @@ unitIdSupported = ghcSupported "Uses unit IDs"
 backpackSupported :: Compiler -> Bool
 backpackSupported = ghcSupported "Support Backpack"
 
+-- | Does this compiler support the -jsem option
```
Missing `?`

ffaf1 commented on Cabal/src/Distribution/Simple/Setup/Build.hs:
```diff
@@ -65,6 +66,7 @@ defaultBuildFlags =
     , buildDistPref = mempty
     , buildVerbosity = Flag normal
     , buildNumJobs = mempty
+    , buildUseSemaphore = NoFlag
```
`NoFlag` or `Serial`?

Kleidukos replied:
Not sure if there's a major practical difference. Let's keep it like that for now.

ffaf1 commented on Cabal/src/Distribution/Simple/Setup/Build.hs:
```diff
@@ -125,6 +127,13 @@ buildOptions progDb showOrParseArgs =
+  , option
+      []
+      ["semaphore"]
+      "semaphore"
```
Could we have a `cabal build --help` to see how this is documented?

Kleidukos replied:
```
 --semaphore                    Use a semaphore so GHC can compile components
                                in parallel
 -j, --jobs[=NUM]               Run NUM jobs simultaneously (or '$ncpus' if no
                                NUM is given).
```

ffaf1 commented on Cabal/src/Distribution/Types/ParStrat.hs:
```diff
@@ -0,0 +1,22 @@
+    Serial
+  deriving (Show)
+
+-- Used by Cabal to indicate that we want to use this specific semaphore (created by cabal-install)
+type ParStrat = ParStratX String
+
+-- Used by cabal-install to say we want to create a semaphore with N slots.
+type ParStratInstall = ParStratX Int
```
Nitpick, missing `|`.

ffaf1 commented on Cabal/src/Distribution/Types/ParStrat.hs:
```diff
@@ -0,0 +1,22 @@
+-- Used by cabal-install to say we want to create a semaphore with N slots.
+type ParStratInstall = ParStratX Int
+
+isParallelBuild :: ParStratX n -> Bool
+isParallelBuild Serial = False
+isParallelBuild NumJobs{} = True
+isParallelBuild UseSem{} = True
```
Missing documentation, and maybe not entirely true? `-j1` is *not* parallel.

ffaf1 commented on cabal-install/src/Distribution/Client/CmdHaddockProject.hs:
```diff
@@ -161,208 +161,216 @@ haddockProjectAction flags _extraArgs globalFlags = do
   -- we need.
   --
 
-  withContextAndSelectors RejectNoTargets Nothing
-                          (commandDefaultFlags CmdBuild.buildCommand)
-                          ["all"] globalFlags HaddockCommand
-                          $ \targetCtx ctx targetSelectors -> do
-    baseCtx <- case targetCtx of
-      ProjectContext -> return ctx
```
I suspect this all is fourmolu and it is *not* good. What has been happening here, @Kleidukos?

ffaf1 commented on cabal-install/src/Distribution/Client/Config.hs:
```diff
@@ -434,6 +434,7 @@ instance Semigroup SavedConfig where
           , installSymlinkBinDir = combine installSymlinkBinDir
           , installPerComponent = combine installPerComponent
           , installNumJobs = combine installNumJobs
+          , installUseSemaphore = combine installUseSemaphore
```
How does this behave with two non-empty values?

Kleidukos replied:
I guess the usual, the more local value takes priority?

ffaf1 commented on cabal-install/src/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -652,57 +656,64 @@ rebuildTargets
-            else newSerialJobControl
+        mkJobControl <- case buildSettingNumJobs of
+          Serial -> newSerialJobControl
+          NumJobs n -> newParallelJobControl (fromMaybe numberOfProcessors n)
```
Same problem, is n=1 case handled correctly?

Kleidukos replied:
Here yes. We could optimise this later when the system is better understood, but for now let's leave it as that.

ffaf1 commented on cabal-install/src/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -652,57 +656,64 @@ rebuildTargets
         createDirectoryIfMissingVerbose verbosity True distTempDirectory
         traverse_ (createPackageDBIfMissing verbosity compiler progdb) packageDBsToUse
 
-        -- Before traversing the install plan, preemptively find all packages that
-        -- will need to be downloaded and start downloading them.
-        asyncDownloadPackages
-          verbosity
-          withRepoCtx
-          installPlan
```
Again fourmolu quizzical edits…

Kleidukos replied:
Not really, due to bracket + do notation, I believe all indentation has to shift to the right.

ffaf1 replied:
the only thing introduced in this pr is a `case`, not a `do`, right?

Kleidukos replied:
Actually it's:

```diff

-        -- Before traversing the install plan, preemptively find all packages that
-        -- will need to be downloaded and start downloading them.
-        asyncDownloadPackages
+        bracket (pure mkJobControl) cleanupJobControl $ \jobControl -> do
+          -- Before traversing the install plan, preemptively find all packages that
+          -- will need to be downloaded and start downloading them.
+          asyncDownloadPackages
```

ffaf1 commented on cabal-testsuite/PackageTests/NewBuild/Semaphore/cabal.test.hs:
```diff
@@ -0,0 +1,6 @@
+    skipUnlessGhcVersion ">= 9.7"
+    -- Parallel flag means output of this test is nondeterministic
+    recordMode DoNotRecord $
+        cabal "v2-build" ["-j", "--semaphore", "Semaphore"]
```
So semaphore is not boolean, but also accepts a name?

Kleidukos replied:
Looking at this screenshot, it would appear that cabal automatically translates the --semaphore option to --semaphore=unique_semaphore_name 

![Screenshot_20230714_110247](https://github.com/haskell/cabal/assets/29253044/bc5db975-b635-46ed-8630-28cb5d649a5a)

Kleidukos commented on changelog.d/pr-8557-2:
```diff
@@ -0,0 +1,16 @@
+packages: Cabal
+prs: #8557
+
+description: {
+
+When `./Setup build --semaphore <SEM>` is called, `ghc` will be called with the
+`-jsem` option. It is the responsibility of the caller of `./Setup build` to
+manage the semaphore according to the GHC Jobserver Protocol.
+
+This low level interface is intended to be called by a high-level tool such as
+`cabal-install` which can create and manage the semaphore appropriately.
```
@ffaf1 this is what you've been seeing: cabal-install handles the uniqueness of the name and uses the more low-level interface of `./Setup build`

Kleidukos commented:
@mergifyIO backport 3.10

ulysses4ever commented:
GitHub needed a rebase and I pushed the button. Results are not hopeful:

- one lint failure
- one actual failure because of a change to the options interface

andreabedini commented:
I understand this was subsumed by #9139 which is now merged. Please reopen if I am mistaken.
